### PR TITLE
Refactor statsbeat

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Changing Statsbeat to non-static. A new instance will be created for every connection string that will be used within a process.

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzureMonitorTransmitter.cs [hunk 1]
```diff
@@ -36,11 +36,31 @@ public AzureMonitorTransmitter(AzureMonitorExporterOptions options, TokenCredent
...
+        }
+
+        private static void InitializeStatsbeat(string? connectionString)
+        {
+            try
+            {
+                // Do not initialize statsbeat for statsbeat.
+                if (connectionString != null && connectionString != Statsbeat.Statsbeat_ConnectionString_EU && connectionString != Statsbeat.Statsbeat_ConnectionString_NonEU)
+                {
+                    _ = new Statsbeat(connectionString);
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzureMonitorTransmitter.cs:57]: Should this method return `StatsBeat?`

If nothing holds a reference to StastBeat, I would expect GC to dispose of this.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs [hunk 2]
```diff
@@ -1,8 +1,6 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-#nullable disable // TODO: remove and fix errors
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs:4]: 🎉

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs [hunk 3]
```diff
@@ -101,6 +97,37 @@ internal static class Statsbeat
...
+
+            s_myMeter.CreateObservableGauge("AttachStatsbeat", () => GetAttachStatsbeat());
+
+            // Configure for attach statsbeat which has collection
+            // schedule of 24 hrs == 86400000 milliseconds.
+            // TODO: Follow up in spec to confirm the behavior
+            // in case if the app exits before 24hrs duration.
+            var exporterOptions = new AzureMonitorExporterOptions();
+            exporterOptions.DisableOfflineStorage = true;
+            exporterOptions.ConnectionString = _statsbeat_ConnectionString;
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs:122]: Are we going to add an `EnablreStatsBeat` to Options?
If so would want to set that here.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzureMonitorTransmitter.cs (hunk 1 above)
Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzureMonitorTransmitter.cs:57]: @vishweshbankwar As we discussed offline, AzureMonitorTransmitter need to implement IDisposable and it should be disposed in exporter. `Statsbeat` object could be disposed in `AzureMonitorTransmitter` dispose.

GC will not dispose this object as it got a child object of MeterProvider instance and that won't get disposed.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs [hunk 4]
```diff
@@ -17,39 +15,37 @@
...
+        private readonly string? _customer_Ikey;
 
-        internal static MeterProvider s_attachStatsBeatMeterProvider;
+        internal MeterProvider? _attachStatsbeatMeterProvider;
 
-        internal static Regex s_endpoint_pattern = new("^https?://(?:www\\.)?([^/.-]+)");
+        internal static Regex s_endpoint_pattern => new("^https?://(?:www\\.)?([^/.-]+)");
 
-        internal static readonly HashSet<string> EU_Endpoints = new()
+        internal static readonly HashSet<string> s_eU_Endpoints = new()
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs:48]: nit: 
```suggestion
        internal static readonly HashSet<string> s_EU_Endpoints = new()
```

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs [hunk 5]
```diff
@@ -119,90 +146,54 @@ private static string GetOS()
...
         {
-            // check whether it is disabled or already initialized.
-            if (s_isEnabled && s_attachStatsBeatMeterProvider == null)
+            var patternMatch = s_endpoint_pattern.Match(ingestionEndpoint);
+            string? statsbeatConnectionString = null;
+            if (patternMatch.Success)
             {
-                if (s_statsBeat_ConnectionString == null)
+                var endpoint = patternMatch.Groups[1].Value;
+                if (s_eU_Endpoints.Contains(endpoint))
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs:156]: nit: same here, try using EU or change the variable name in line 160 to  match this format

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs [hunk 6]
```diff
@@ -222,17 +213,17 @@ private static VmMetadataResponse GetVmMetadataResponse()
             }
         }
 
-        private static void SetResourceProviderDetails()
+        private void SetResourceProviderDetails()
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs:216]: Later, we might need to rely on resource detector to get the details.

**Review by rajkumar-rangaraj — APPROVED** [review]
Left few minor comments and dispose could be addressed in a follow up PR.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzureMonitorTransmitter.cs [hunk 7]
```diff
@@ -36,11 +36,31 @@ public AzureMonitorTransmitter(AzureMonitorExporterOptions options, TokenCredent
 
             options.Retry.MaxRetries = 0;
 
-            _connectionVars = InitializeConnectionVars(options);
+            var connectionString = GetConnectionString(options);
+            _connectionVars = ConnectionStringParser.GetValues(connectionString);
```
[the remark this answers is not in the document]
Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzureMonitorTransmitter.cs:40]: thinking about this some more. Do you need the full connection string?
I think you can do everything with the Ikey and Endpoint parts.

[the remark this answers is not in the document]
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzureMonitorTransmitter.cs]: Fixed

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzureMonitorTransmitter.cs (hunk 1 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzureMonitorTransmitter.cs]: Added todo here

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs (hunk 3 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs:120]: This is something I am not sure about. We would have something to disable it. But I don't know if it should be part of public options. Spec suggests using env plus appsettings. (Lets discuss first)

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs [hunk 8]
```diff
@@ -101,6 +97,35 @@ internal static class Statsbeat
             "westus3",
         };
 
+        internal Statsbeat(ConnectionVars? connectionStringVars)
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/Statsbeat/Statsbeat.cs:100]: ConnectionVars shouldn't be null :)

```suggestion
        internal Statsbeat(ConnectionVars connectionStringVars)
```

**Review by TimothyMothra — APPROVED** [review]
LGTM!
left one minor comment

**Pull request merged**